### PR TITLE
Updated cuda exchange to use 3N matrix. Added cuda stats calculation …

### PR DESCRIPTION
**Pull request by mattoaellis**
…for 1 material type.

### CODE REVIEW

File: src/cuda/statistics.cu [hunk 1]
```diff
@@ -449,11 +491,11 @@ namespace vcuda{
                   hist[4 * i + 0] = mx * imm;
                   hist[4 * i + 1] = my * imm;
                   hist[4 * i + 2] = mz * imm;
-                  hist[4 * i + 3] = imm * ms;
+                  hist[4 * i + 3] = ms / imm;
```
Comment (mattoaellis) [inline src/cuda/statistics.cu]: imm is actuall 1/|M|. Actually looking at it again this probably needs to be 1.0/( ms* imm)

**Comment by richard-evans** [comment]
Thanks Matt. Quick question - why is 3N matrix forced all the time - presumably a 1N matrix is faster for isotropic interactions? Also, don't we still need the case statement for the tensorial exchange?

**Comment by mattoaellis** [comment]
Hi Richard,
The idea about using the 3N matrix over the 1N is that in CUDA it is generally more efficient to have 1 large operation as opposed to multiple little ones since there is a kernel launch overhead. I think this will be true especially for the tensorial exchange which would be 9 times the kernel overhead. IIRC I did test it and found even with the extra memory copies for the 1N to 3N arrays and back it was slightly faster. The memory copies should be pretty fast as it is just loading then storing with aligned data accesses. Also the 3N matrix then has the advantage of incorporating vector and tensor without having to declare more variables. I had thought I had done the tensorial exchange but it seems I hadn't, which needs to go in the case statement when the matrix is initialised. We shouldn't need a case statement later on since all the information is already encoded into the array and since it is sparse format won't be doing any extra calculations.

Do you want to merge this now and then I can work on the tensor exchange or shall I do that then issue a new pull request?

**Comment by richard-evans** [comment]
Hi Matt,

Thanks - that all makes sense! If you have time to implement the exchange tensor that would be great! I can then merge this in and add a few other bits and pieces such as the new dipole-dipole and also atomistic dipole-dipole calculations.

Cheers,

Richard

**Comment by mattoaellis** [comment]
Hi Richard,

I have added in the code for the tensor exchange now if you want to merge it into yours. It compiles but I don't have an input file to test the tensor exchange specifically, if you have anything handy could you send it me then I can test it out fully. There is still a bug when the code exits and some array is not properly deallocated but I couldn't find which one it is.

Cheers,
Matt

**Comment by richard-evans** [comment]
Hi Matt,

Thanks very much! I have attached example files with tensorial exchange/dmi (built internally) which I think should show stripe domains. I will see if I can track down the bug.

Cheers,

Richard

[Co.txt](https://github.com/richard-evans/vampire/files/3084650/Co.txt)
[input.txt](https://github.com/richard-evans/vampire/files/3084651/input.txt)

**Comment by mattoaellis** [comment]
Thanks Richard.
I tried out testing the dmi but the problem was the cuda branch is forked off an older version of Vampire without the dmi functionality. I've pulled the latest master branch and 'merged' it into my cuda branch. I had to the add in a few changes to cuda exchanges since the exchange_type is now handled differently but now the test isn't running properly so I'll try to track down what is causing that.

**Comment by mattoaellis** [comment]
I have tweaked the cuda tensor exchange and it appears to be working now for both a timeseries and DMI test. Attached is the output using the DMI for both serial and cuda implementations, I guess any difference is due to the thermalisation at the start. I think you might be okay to merge my branch into yours now. 

Screenshot 2019-04-17 at 10 36 21

**Pull request merged**